program-runtime/src/loaded_programs.rs

```diff
@@ -367,13 +367,32 @@ pub struct LoadedProgramsForTxBatch {
     /// LoadedProgram is the corresponding program entry valid for the slot in which a transaction is being executed.
     entries: HashMap<Pubkey, Arc<LoadedProgram>>,
     slot: Slot,
+    /// Runtime environment reference from LoadedProgram cache
+    pub program_runtime_environment_v1: Arc<BuiltinProgram<InvokeContext<'static>>>,
+    /// RuntimeV2 environment reference from LoadedProgram cache
+    pub program_runtime_environment_v2: Arc<BuiltinProgram<InvokeContext<'static>>>,
 }
 
 impl LoadedProgramsForTxBatch {
-    pub fn new(slot: Slot) -> Self {
+    pub fn new_from_cache(slot: Slot, loaded_programs_cache: &LoadedPrograms) -> Self {
```

Comment:
Can you factor out `program_runtime_environment_v1` and `program_runtime_environment_v2` into its own `ProgramRuntimeEnvironments` struct? These will become 4 in total because each gets its "upcoming" variant in #32172.

Then we can just pass that as a parameter here and only need one constructor. Also, if by implementing `Clone` for `ProgramRuntimeEnvironments` we can avoid cloning all individually here and in other places.